pkg/nsx/services/realizestate/realize_state.go

```diff
@@ -0,0 +1,47 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package realizestate
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+)
+
+type RealizeStateService struct {
+	common.Service
+}
+
+func InitializeRealizeState(service common.Service) *RealizeStateService {
+	return &RealizeStateService{
+		Service: service,
+	}
+}
+
+// CheckRealizeState allows the caller to check realize status of entityType with retries.
+// backoff defines the maximum retries and the wait interval between two retries.
+func (service *RealizeStateService) CheckRealizeState(backoff wait.Backoff, orgID, projectID, intentPath, entityType string) error {
+	return retry.OnError(backoff, func(err error) bool {
+		return err != nil
+	}, func() error {
+		results, err := service.NSXClient.RealizedStateClient.List(orgID, projectID, intentPath, nil)
+		if err != nil {
+			return err
+		}
+		for _, result := range results.Results {
+			if *result.EntityType != entityType {
+				continue
+			}
+			if *result.State == "REALIZED" {
+				return nil
+			}
+			return errors.New(fmt.Sprintf("%s not realized", entityType))
+		}
+		return errors.New(fmt.Sprintf("%s not realized", entityType))
```

Comment:
Can we return the original error retrieved at L32? E.g. we need to delete the subnet port in GenericPolicyRealizedResource_STATE_ERROR but don't need to delete it in the other unrealized state.

Reply:
The original error is returned in L34?

Reply:
> Can we return the original error retrieved at L32? E.g. we need to delete the subnet port in GenericPolicyRealizedResource_STATE_ERROR but don't need to delete it in the other unrealized state.

Got your point, but I found that `GenericPolicyRealizedResource_STATE_ERROR` is string type(not error type), so it can't be used directly. So how about returning errors.New(*result.state) and using `error.Error() == GenericPolicyRealizedResource_STATE_ERROR` to make the judgement?

Reply:
> Can we return the original error retrieved at L32? E.g. we need to delete the subnet port in GenericPolicyRealizedResource_STATE_ERROR but don't need to delete it in the other unrealized state.

Got your point, but I found that `GenericPolicyRealizedResource_STATE_ERROR` is string type(not error type), so it can't be used directly. So how about returning errors.New(*result.state) and using `error.Error() == GenericPolicyRealizedResource_STATE_ERROR` to make the judgement?